File: 1. Creational Design Patterns/Factory Method/Javascript/FactoryMethodExample.js
// Factory Method – JavaScript Example
class Transport {
  deliver() {}
}

class Truck extends Transport {
  deliver() {
    console.log('Deliver by land in a box.');
  }
}

class Ship extends Transport {
  deliver() {
    console.log('Deliver by sea in a container.');
  }
}

class Logistics {
  planDelivery() {
    const t = this.createTransport();
    t.deliver();
  }

  createTransport() {}
}

class RoadLogistics extends Logistics {
  createTransport() {
    return new Truck();
  }
}

class SeaLogistics extends Logistics {
  createTransport() {
    return new Ship();
  }
}

const logistics = process.argv[2] === 'sea'
  ? new SeaLogistics()
  : new RoadLogistics();

logistics.planDelivery();

File: 1. Creational Design Patterns/Factory Method/Kotlin/FactoryMethodExample.kt
// Factory Method – Kotlin Example
interface Transport {
    fun deliver()
}

class Truck : Transport {
    override fun deliver() = println("Deliver by land in a box.")
}

class Ship : Transport {
    override fun deliver() = println("Deliver by sea in a container.")
}

abstract class Logistics {
    fun planDelivery() {
        createTransport().deliver()
    }
    protected abstract fun createTransport(): Transport
}

class RoadLogistics : Logistics() {
    override fun createTransport() = Truck()
}

class SeaLogistics : Logistics() {
    override fun createTransport() = Ship()
}

fun main(args: Array<String>) {
    val l: Logistics = if (args.isNotEmpty() && args[0] == "sea") {
        SeaLogistics()
    } else {
        RoadLogistics()
    }
    l.planDelivery()
}
